### PR TITLE
[WIP] [refactor] Move logger.finalize to Trainer's teardown call

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This is currently only being called at the end of the fit loop. However, this would be useful in `validate`/`test`/ `predict` too.
As a result, can we move this to the Trainer's `teardown` to centralize this?
pros:
- less loop-specific handling (we might forget to do this)

cons:
- ?

Not yet ready for review: checking CI

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
Hey! I like the idea and for single processes this works great. 

However the concern for **ddp spawn** is the following:
`on_run_end()`, where `finalize()` was originally, happens inside the spawned process where the logger is active. 
Now you moved it the call to after dispatch which means that this finalize() will happen in the master process where no training/logging ever happened. I need to find a better place where to move this.

**Comment by ananthsub** [comment]
@awaelchli I see, so then the question is:
- Do we repeat this across loops, similar to how `profiler.describe` is handled today? However, as seen here, we can forget this in loop implementations, and with fit and evaluation loops, we need to add a check for whether we're in trainer.fit vs trainer.test/trainer.validate
- Or do we special case the spawn plugins and handle finalization/teardown in post_dispatch?  

Neither sounds too great :/

**Comment by ananthsub** [comment]
@awaelchli n00b question: would calling `trainer._call_teardown_hook` here work? https://github.com/PyTorchLightning/pytorch-lightning/blob/1afc1ca7ef8d069895f4065fb4000b24771585f4/pytorch_lightning/plugins/training_type/ddp_spawn.py#L205-L208

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1114,6 +1114,12 @@ def _call_teardown_hook(self, model: 'pl.LightningModule') -> None:
         self.teardown(stage=fn)
         model.teardown(stage=fn)
 
+        # todo: TPU 8 cores hangs in flush with TensorBoard. Might do for all loggers.
+        # It might be related to xla tensors blocked when moving the cpu
+        # kill loggers
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1119]: @kaushikb11 Does this sound good with you ?

**Comment by ananthsub** [comment]
This was taken care of in https://github.com/PyTorchLightning/pytorch-lightning/pull/8685

**Pull request closed without merging**